### ISSUE TITLE
About Training Dataset in BLIPo3-Next

### ISSUE THREAD

**Issue by Columbine21** [JiuhaiChen/BLIP3o#78]
https://github.com/JiuhaiChen/BLIP3o/blob/6ced364eeee8b2d53e9491ffae48b00a0271cda2/blip3o/data/dataset.py#L175

Hi, Thanks for the great work~

The question is How the training dataset is obtained in the BLIP3o-Next Training Dataset? How to obtained the "/fsx/home/jiuhai.chen/soda/overfit.tar" file used in the training scripts?

Thanks a lot!

**Comment by Columbine21** [comment]
https://github.com/JiuhaiChen/BLIP3o/blob/6ced364eeee8b2d53e9491ffae48b00a0271cda2/blip3o/model/language_model/blip3o_qwen.py#L146

Besides, there may be a bug here (hidden_states[b, start:end, :] returns a 2-d tensor, instead of 3-d tensor)

**Comment by JiuhaiChen** [comment]
Hi, thanks for your interest. We use BLIP3o data which can be downloaded from the huggingface space, and replace the data path in the dataset.py

 **Fully Open-Source:**
  - **Pretraining Data:** [27 Million Detailed Captions](https://huggingface.co/datasets/BLIP3o/BLIP3o-Pretrain-Long-Caption), [5 Million Short Captions](https://huggingface.co/datasets/BLIP3o/BLIP3o-Pretrain-Short-Caption)
  - **Instruction Tuning Data:** [BLIP3o-60k](https://huggingface.co/datasets/BLIP3o/BLIP3o-60k), [ShareGPT-4o-Image](https://huggingface.co/datasets/FreedomIntelligence/ShareGPT-4o-Image)
  - **Model Weights (3B):** [Pretrain](https://huggingface.co/BLIP3o/BLIP3o-NEXT-Pretrain), [Instruction Tuning](https://huggingface.co/BLIP3o/BLIP3o-NEXT-SFT), [GRPO-Geneval](https://huggingface.co/BLIP3o/BLIP3o-NEXT-GRPO-Geneval), [GRPO-Text]()
  - **Training Code:** Pretrain, Instruction Tuning, GRPO


Yes, thanks for the reminder. hidden_states_filter is a 2D tensor with shape [num_tokens, hidden_dim], but it seems fine to access hidden_states[b, start:end, :] since we're not modifying the third dimension.

Since each image is represented by 730 tokens, we want to ensure that we have exactly 730 image tokens. If not, we just take the last 730 tokens to guarantee the correct number. Let me know of you have more questions.

**Comment by Columbine21** [comment]
@JiuhaiChen Thanks for your fast reply! 

According to your reply, the '/fsx/home/jiuhai.chen/soda/overfit.tar' contains all .tar files in  BLIP3o-Pretrain-JourneyDB, BLIP3o-Pretrain-Long-Caption, and BLIP3o-Pretrain-Short-Caption? 

It would be helpful to provide how to obtain the "/fsx/home/jiuhai.chen/soda/overfit.tar (specifically, which folder / files does it contains in detail?)

**Comment by Columbine21** [comment]
Besides, num_proc=1 leads to a lot of time performing data preprocessing. However, num_proc>1 will lead to Bugs in my own environment (is this why here requires num_proc=1? )

**Comment by JiuhaiChen** [comment]
Hi, you'll need to download all the datasets and update the path '/fsx/home/jiuhai.chen/soda/overfit.tar' to point to your own directory that contains all the .tar files. You can refer to the Hugging Face WebDataset documentation for guidance on loading the datasets.

Also, I encountered a bug when setting num_proc > 1, but this only affects the first run. Once the cache files are generated, you can reuse them in future runs without issues.

**Comment by Columbine21** [comment]
> Hi, thanks for your interest. We use BLIP3o data which can be downloaded from the huggingface space, and replace the data path in the dataset.py
> 
> **Fully Open-Source:**
> 
> * **Pretraining Data:** [27 Million Detailed Captions](https://huggingface.co/datasets/BLIP3o/BLIP3o-Pretrain-Long-Caption), [5 Million Short Captions](https://huggingface.co/datasets/BLIP3o/BLIP3o-Pretrain-Short-Caption)
> * **Instruction Tuning Data:** [BLIP3o-60k](https://huggingface.co/datasets/BLIP3o/BLIP3o-60k), [ShareGPT-4o-Image](https://huggingface.co/datasets/FreedomIntelligence/ShareGPT-4o-Image)
> * **Model Weights (3B):** [Pretrain](https://huggingface.co/BLIP3o/BLIP3o-NEXT-Pretrain), [Instruction Tuning](https://huggingface.co/BLIP3o/BLIP3o-NEXT-SFT), [GRPO-Geneval](https://huggingface.co/BLIP3o/BLIP3o-NEXT-GRPO-Geneval), GRPO-Text
> * **Training Code:** Pretrain, Instruction Tuning, GRPO
> 
> Yes, thanks for the reminder. hidden_states_filter is a 2D tensor with shape [num_tokens, hidden_dim], but it seems fine to access hidden_states[b, start:end, :] since we're not modifying the third dimension.
> 
> Since each image is represented by 730 tokens, we want to ensure that we have exactly 730 image tokens. If not, we just take the last 730 tokens to guarantee the correct number. Let me know of you have more questions.

@JiuhaiChen Thanks for your replay~ 

According to the paper, **Pretraining Data** also contain the BLIP3o-Pretrain-JourneyDB? so Pretraining Data is  [27 Million Detailed Captions](https://huggingface.co/datasets/BLIP3o/BLIP3o-Pretrain-Long-Caption), [5 Million Short Captions](https://huggingface.co/datasets/BLIP3o/BLIP3o-Pretrain-Short-Caption) + 4 Million JourneyDB data?

**Comment by JiuhaiChen** [comment]
Yeah, exactly.

**Comment by Columbine21** [comment]
@JiuhaiChen Thanks for your reply~

I would like to inquire, during the BLIP3o-Next Pretrain stage, if it would be more straightforward to only use the next token prediction task without utilizing diffusion loss, and employ the predicted VQ tokens for image generation (thus avoiding the additional parameters introduced by the diffusion model for image generation). Alternatively, I would like to know if the impact of diffusion loss on the pretraining stage of the model is significant, and whether it is feasible to conduct the first stage of training solely using NTP. 

BTW, all tokens (include the user's query are used for CELoss supervision), which is different from the common way (only assistant token are used for celoss computation.) How was this design conceived? Was it intended to preserve the comprehension ability of the language model?

The last question is about the pretraining phase results. Could the author provide an overview of the training loss curve during the pretrain/sft phase, or roughly estimate the range of the final training loss (diffusion loss/CE loss) values for each stage.

Thanks for your great work and help

**Comment by JiuhaiChen** [comment]
BLIP3o-NEXT still relies on a diffusion model to generate the final image, similar to MetaQuery and BLIP3o. The key difference is that we take more benefits of autoregressive model. Instead of generating the learnable query in a single forward pass (as in BLIP3o and MetaQuery), we use autoregressive style to generate the condition to diffusion model. This sequential design offers benefits for tasks such as text rendering (will put some examples in the coming paper), and the discrete token is more compatible with RL frameworks.

I believe the architecture of unified model will ultimately converge to a framework where an autoregressive model generates the condition (whether discrete or continuous), and the diffusion model then produces the final images conditioned on it. However, the open questions remain: how should the condition be generated, and how should it be trained? (For example AR and diffusion model are jointly optimized?) Both BLIP3o and BLIP3o-NEXT are views as proof of concept explorations toward this direction.


The language capability is not maintained during the pretraining stage. However, if some text corpus data is added, the language capability can be preserved.


The training loss: https://api.wandb.ai/links/jchen169/kb1n4no8

**Comment by Columbine21** [comment]
Thanks for your help~
